### PR TITLE
cmd/snap-confine: be compatible with a snap rootfs built as a tmpfs

### PR DESCRIPTION
**Pull request by mardy**
In order to support home directories at arbitrary locations, https://github.com/snapcore/snapd/pull/11367 was created to build the snap rootfs as a tmpfs, on which the needed directories from the base snap and from the host are mounted. However, one problem with that approach is that older versions of snapd won't support it, and any running snaps would break if snapd is rolled back from a version supporting a tmpfs rootfs to a version which doesn't.

To mitigate this problem, we now introduce an intermediate step: we do not alter the way that snap-confine generates the snap filesystem, but make it (and snap-update-ns) compatible with a future version of snapd which might build the snap rootfs as a tmpfs. In this way, we allow such a future version of snapd to be rolled back to the current version of snapd without breaking the running applications. Of course, a rollback to an even older version of snapd (as that could happen when rolling back from one LTS to the previous one) would still cause breakage, but that is a shortcoming we could tolerate (or we could force a reboot after a rollback to such an older snapd).

### CODE REVIEW

File: cmd/snap-update-ns/change.go [hunk 1]
```diff
@@ -543,6 +543,13 @@ func neededChanges(currentProfile, desiredProfile *osutil.MountProfile) []*Chang
 		}
 		skipDir = "" // reset skip prefix as it no longer applies
 
+		if current[i].XSnapdOrigin() == "rootfs" {
+			// This is the rootfs setup by snap-confine, we should not touch it
+			logger.Debugf("reusing rootfs")
+			reuse[dir] = true
+			continue
+		}
+
```
Comment (pedronis) [inline cmd/snap-update-ns/change.go:552]: this needs a unit test

**Review by pedronis — COMMENTED** [review]
lgtm, test remark

File: cmd/snap-update-ns/change.go (hunk 1 above)
Comment (mardy) [reply cmd/snap-update-ns/change.go:552]: Added!

**Review by pedronis — DISMISSED** [review]
thanks

File: cmd/snap-confine/ns-support.c [hunk 2]
```diff
@@ -289,7 +289,14 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
...
 	     mie = sc_next_mountinfo_entry(mie)) {
-		if (!sc_streq(mie->mount_dir, "/")) {
+		// If the "/" directory is a tmpfs, then we'll look at the /usr
+		// directory
+		if (sc_streq(mie->fs_type, "tmpfs")) {
+			continue;
+		}
+
+		if (!sc_streq(mie->mount_dir, "/") &&
+			!sc_streq(mie->mount_dir, "/usr")) {
```
Comment (bboozzoo) [inline cmd/snap-confine/ns-support.c:299]: is this check still correct if the snap layout adds a new directory under /usr? just modified test-snapd-layout to add a layout entry under `/usr/foo`, which then makes /usr be reported as tmpfs:
```
$ stat -f /usr
  File: "/usr"
    ID: 6585fe8765d1a97e Namelen: 255     Type: tmpfs
Block size: 4096       Fundamental block size: 4096
Blocks: Total: 4100692    Free: 4100692    Available: 4100692
Inodes: Total: 4100692    Free: 4100681
```

As far as path prefix goes, the layout was not allowed to create top level directories only. Perhaps we could look at `/writable` as it makes no sense to modify that one with layouts?

Comment (mardy) [reply cmd/snap-confine/ns-support.c:299]: Thanks Maciek, you got a point there! I've prepared a change to deal with this (https://paste.ubuntu.com/p/rMtBdTfm5K/), but then while testing it (adding a layout for `/usr/foo`) I noticed that when the /usr tmpfs mimic is created, the original /usr device is still listed in mountinfo:
```
...
631 593 7:1 / /var/lib/snapd/hostfs/snap/core20/1518 ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
...
649 592 0:53 / / rw,relatime - tmpfs none rw,mode=770
650 649 7:1 /usr /usr ro,nodev,relatime master:396 - squashfs /dev/loop1 ro     <--- THIS LINE
...
617 650 0:54 / /usr rw,relatime - tmpfs tmpfs rw,mode=755
618 617 7:1 /usr/bin /usr/bin ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
619 617 7:1 /usr/games /usr/games ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
620 617 7:1 /usr/include /usr/include ro,nodev,relatime master:396 - squashfs /dev/loop1 ro
...
615 617 7:2 /bin/cmd /usr/foo ro,nodev,relatime master:404 - squashfs /dev/loop2 ro
```

So the branch currently pushed here should handle the situation correctly. Though I'm not sure we want to rely on the existance of that `/usr` -> `/usr` bind mount, as it's a bit too subtle IMHO. So, if you agree, I'll push the change from the pastebin.

Comment (bboozzoo) [reply cmd/snap-confine/ns-support.c:299]: The diff looks good. Please take a look at the formatting, I think `make fmt` should do the trick.

File: cmd/snap-confine/ns-support.c [hunk 3]
```diff
@@ -287,20 +287,38 @@ static bool should_discard_current_ns(dev_t base_snap_dev)
...
+	 *    with all of the original directory entries inside of /usr being
+	 *    bind-mounted onto mount-points created into the tmpfs.
+	 * In light of the above, we do ignore all tmpfs entries and accept that
+	 * our base snap might be mounted under /, /usr, or anywhere under /usr.
+	 */
 	for (mie = sc_first_mountinfo_entry(mi); mie != NULL;
 	     mie = sc_next_mountinfo_entry(mie)) {
-		if (!sc_streq(mie->mount_dir, "/")) {
+		// If the "/" directory is a tmpfs, then we'll look at the /usr
+		// directory
```
Comment (bboozzoo) [inline cmd/snap-confine/ns-support.c:305]: the comment is probably no longer needed
```suggestion
```

**Review by bboozzoo — APPROVED** [review]
LGTM

File: cmd/snap-confine/ns-support.c (hunk 3 above)
Comment (mardy) [reply cmd/snap-confine/ns-support.c]: Oops, removed!

**Review by pedronis — APPROVED** [review]
lgtm to me, but the change is involved enough now that this needs to be looked at by security as well

**Review by alexmurray — APPROVED** [review]
LGTM!

**Comment by mardy** [comment]
@pedronis, @alexmurray: sorry to ping you again, but the spread tests revealed a critical (and extremely silly!) error that was causing snap-confine to crash in some cases. I added a commit that fixes that; I don't think it invalidates your reviews, but just FYI: https://github.com/snapcore/snapd/pull/11665/commits/dca8e2c806ff51bac4baaa6a8da1b3cf053a73f3

**Comment by alexmurray** [comment]
Thanks for the @mardy - apologies I should have picked that up (I just ran Coverity over your PR before this new change and it flagged that as a potential issue so I should have ran it when doing the original PR review).

**Comment by mardy** [comment]
I added one commit to exclude the 32bit version of ubuntu 18.04 from the spread test, since in the test I'm using core20 which is not available in 32bit. I gave a quick try to change the test snap to use core18 but then I got an error (umount command not found), which I guess is fixable, but I'm not sure it's worth it.

Please confirm if the PR is fine as is (it's green, for once! :-D) or if you'd like me to revert the ubuntu-18.04-32 exclusion and improve the tests instead.

**Comment by mvo5** [comment]
This looks fine to me and it has needed reviews, I will merge.

**Pull request merged**